File: src/main/java/komu/blunt/core/CoreDefineExpression.java
package komu.blunt.core;

import komu.blunt.analyzer.VariableReference;
import komu.blunt.asm.Assembler;
import komu.blunt.asm.Instructions;
import komu.blunt.asm.Linkage;
import komu.blunt.asm.Register;
import komu.blunt.stdlib.BasicValues;

import static com.google.common.base.Preconditions.checkNotNull;

public final class CoreDefineExpression extends CoreExpression {

    private final CoreExpression expression;
    private VariableReference var;

    public CoreDefineExpression(CoreExpression expression, VariableReference var) {
        this.expression = checkNotNull(expression);
        this.var = checkNotNull(var);
    }

    @Override
    public Instructions assemble(Assembler asm, Register target, Linkage linkage) {
        Instructions instructions = new Instructions();
        
        instructions.append(expression.assemble(asm, target, Linkage.NEXT));

        instructions.storeVariable(var, target);

        instructions.loadConstant(target, BasicValues.UNIT);
        instructions.finishWithLinkage(linkage);

        return instructions;
    }
}

File: src/main/kotlin/komu/blunt/core/CoreDefineExpression.kt
package komu.blunt.core

import komu.blunt.analyzer.VariableReference
import komu.blunt.asm.*
import komu.blunt.stdlib.BasicValues

class CoreDefineExpression(private val variable: VariableReference, private val expression: CoreExpression) : CoreExpression() {

    override fun assemble(asm: Assembler, target: Register, linkage: Linkage) =
        instructions {
            instructionsOf(expression.assemble(asm, target, Linkage.Next))
            storeVariable(variable, target)

            loadConstant(target, BasicValues.UNIT)
            finishWithLinkage(linkage)
        }

    override fun simplify() = CoreDefineExpression(variable, expression.simplify())

    override fun toString() = "(define $variable $expression)"
}

